fix: GeneralizedLibrary does not respect inputs_per_library with three or more libraries (#540)

ChengYuHan0406 opened the pull request:
This commit fixes the incorrect handling of `inputs_per_library` when tensoring more than two libraries.

Specifically, it modifies `BaseFeatureLibrary.__mul__()` (triggered by `np.prod` in `GeneralizedLibrary.fit()`) and makes corresponding updates to `TensoredLibrary.get_feature_names()` and `TensoredLibrary.transform()`.

---

### Original behavior

The original implementation of `__mul__()` was:

```python
def __mul__(self, other):
    return TensoredLibrary([self, other])
```

For two libraries (e.g. `[lib_a, lib_b]`), this works correctly:

- `np.prod([lib_a, lib_b])` calls `__mul__()` once.

- Both `self` and `other` are `CustomLibrary` instances.

The returned `TensoredLibrary.libraries` is [`CustomLibrary`, `CustomLibrary`], which is consistent.

However, with three libraries (e.g. `[lib_a, lib_b, lib_c]`):

- `__mul__()` is called twice.

- In the second call, `self` is already a `TensoredLibrary` and other is a `CustomLibrary`.

The returned `TensoredLibrary.libraries` becomes [`TensoredLibrary`, `CustomLibrary`] — always a length-2 list, regardless of the actual number of libraries.

As a result, when the number of libraries > 2, the length of `inputs_per_library` no longer matches the flattened list of `libraries`. This inconsistency causes incorrect behavior in `TensoredLibrary.get_feature_names()` and `TensoredLibrary.transform()`.

---

### Fix

This commit modifies `__mul__()` to ensure proper flattening:

```python
def __mul__(self, other):
    if isinstance(self, TensoredLibrary):
        return TensoredLibrary(self.libraries + [other])
    else:
        return TensoredLibrary([self, other])
```

With this change, the `TensoredLibrary` returned by `np.prod([lib_1, lib_2, ..., lib_n])` will always have libraries of length `n`, consistent with `inputs_per_library`.

In addition, `TensoredLibrary.get_feature_names()` and `TensoredLibrary.transform()` were updated accordingly to correctly handle the flattened list of `libraries`.

Jacob-Stevens-Haas commented on pysindy/feature_library/base.py:
```diff
@@ -451,7 +454,7 @@ def transform(self, x_full):
+            xp = None
             for i in range(len(self.libraries)):
                 lib_i = self.libraries[i]
                 if self.inputs_per_library is None:
```
`self.inputs_per_library` is always set in [`TensoredLibrary.fit()`](https://github.com/dynamicslab/pysindy/blob/c1da1f9e25b008e2f3016f2462ce100d739ea3f9/pysindy/feature_library/base.py#L411), and the function begins with ` check_is_fitted`.  So this conditional isn't necessary.

Jacob-Stevens-Haas commented on pysindy/feature_library/base.py:
```diff
@@ -451,7 +454,7 @@ def transform(self, x_full):
         xp_full = []
         for x in x_full:
-            xp = []
+            xp = None
```
You currently set `xp` to `None` and then handle the first library as a special case at the bottom of the loop (i.e. `if xp is None`).  Instead, start by setting `xp` to the first library's transform.  Then, begin your loop at the second index (`range(1, len(self.libraries)`).

I'd prefer this because (a) removes a conditional/level of indentation and (b) `type(xp)` is always `AxesArray`, which helps static type checking.

Also, please see my next comment for something that will make this easier.

Jacob-Stevens-Haas commented on pysindy/feature_library/base.py:
```diff
@@ -499,20 +497,12 @@ def get_feature_names(self, input_features=None):
-                )
+
+            if len(feature_names) == 0:
+                feature_names = lib_i_feat_names
+            else:
+                feature_names = self._name_combinations(feature_names, lib_i_feat_names)
+
```
Similar to above comments:
1. Love how you removed the nested iteration!  Thank you.
2. Set `input features` once above loop instead of checking if it's `None` at each iteration and setting `input_features_i`.
3. Initialize `feature_names` with the result of applying the first library, then iterate through the rest of the libraries, as above.
4. Could you please add static typing to this function signature?
```
    def get_feature_names(self, input_features: list[str] | None=None) -> list[str]:
```

Jacob-Stevens-Haas reviewed (changes requested):
Thank you for this!  My main ask is that you add a test for #540 to test/test_feature_library.py.  It should fail on current master, but pass on your branch.

I really like this, especially cleaning up the nested loops.  I have a few comments to support better static typing and fewer indentations.  Should be a quick fix, but the main thing is adding a test/tests

ChengYuHan0406 commented:
Hi @Jacob-Stevens-Haas, thanks a lot for the review!

I’ve updated the PR with a couple of changes:

- Before my original commit: I added a test for `TensoredLibrary` that fails on the current master branch.

- After my original commit: I refactored `TensoredLibrary.transform()` and `TensoredLibrary.get_feature_names()` based on your feedback.

Regarding your suggestion to remove the `if xp is None:` check by initializing `xp` with the result of the first library and then looping from index 1:

The main point to note is the logic for handling the default cases:

```python
# TensoredLibrary.transform()

...

lib_i = self.libraries[i]
if self.inputs_per_library is None:
    xp_i = lib_i.transform([x])[0]
else:
    xp_i = lib_i.transform(
        [x[..., _unique(self.inputs_per_library[i])]]
    )[0]

...
```

and

```python
# TensoredLibrary.get_feature_names()

...

lib_i = self.libraries[i]
if input_features is None:
    input_features_i = [
        "x%d" % k for k in _unique(self.inputs_per_library[i])
    ]
else:
    input_features_i = np.asarray(input_features)[
        _unique(self.inputs_per_library[i])
    ].tolist()
lib_i_feat_names = lib_i.get_feature_names(input_features_i)

...
```

are also needed for the first library.

If we initialize `xp` outside the loop, this block would need to be duplicated both outside the loop (for the first library) and inside the loop (for the remaining libraries).

To avoid this redundancy, I extracted the shared logic into two helper methods, `_transform_at()` and `_feature_names_at()`, which are then reused in `transform()` and `get_feature_names()`, respectively.

Does this approach look reasonable to you? Happy to make further adjustments if you think there’s a cleaner way. Thanks again!

Jacob-Stevens-Haas commented on pysindy/feature_library/base.py:
```diff
@@ -487,32 +509,11 @@ def get_feature_names(self, input_features=None):
-                feature_names += self._name_combinations(
-                    lib_i_feat_names, lib_j_feat_names
-                )
+        feature_names = self._feature_names_at(0, input_features)
```
```suggestion
        check_is_fitted(self)

        if input_features is None:
            input_features = ["x%d" % i for i in range(self.n_features_in_)]
        input_features = np.array(input_features)
        inputs0 = input_features[self.inputs_per_library[0]]
        feature_names = self.libraries[0].get_feature_names(inputs0)
```

It's ok to require `get_feature_names()` be called after `fit()`, since the purpose of `fit()` for function libraries is just to determine `self.n_features_in_` and `self.n_features_out_`. `CustomLibrary` and several others already do this.

Jacob-Stevens-Haas commented on pysindy/feature_library/base.py:
```diff
@@ -487,32 +509,11 @@ def get_feature_names(self, input_features=None):
-                )
+        feature_names = self._feature_names_at(0, input_features)
+        for i in range(1, len(self.libraries)):
+            lib_i_feat_names = self._feature_names_at(i, input_features)
```
same as for library 0, we don't need `self._feature_names_at()` since we no longer have to worry about `input_features` ever being `None`

Jacob-Stevens-Haas reviewed (changes requested):
Thanks for adding the tests!  Do all of the cases test something different? Or does a single, simple case with two or three libraries cover the bug?

> The main point to note is the logic for handling the default cases

Exactly!  But that's why I mentioned that `self.inputs_per_library` is never `None`.  Since you don't need the conditional check, `_transform_at()` would just become a single line or two-line function, so you could get rid of the helper function.

There is something similar for `get_feature_names()`.  Since it's always run after `fit()` (add a `check_is_fitted(self)`), you can rely on `self.n_features_in_`.  If you use this to set `input_features` to something non-none, you don't need to check each iteration.  Likewise, `_feature_names_at()` becomes a one or two line function, so you can get rid of the helper function.

Looks like tests and builds fail on 3.9.  The `|` type operator isn't supported in 3.9, but we don't need to support 3.9 anymore.  I'll bump the CI versions.

ChengYuHan0406 commented:
>Exactly! But that's why I mentioned that `self.inputs_per_library` is never `None`

Oh, I see — I completely missed that point earlier. Thanks for emphasizing it again!

I’ve removed the helper functions and inlined the necessary parts back into `transform()` and `get_feature_names()`, and tested it locally.

Regarding the test cases: they all check the same bug but with different numbers of libraries being tensored. The issue can be reproduced on the current master branch whenever the number of libraries > 2 (e.g., `tensor_array = [1, 1, 1, 0, 0]` in `pytest.mark.parametrize`).

Would you prefer that I keep only the simplest case in the test, or keep multiple cases for completeness?
